cpp/oneapi/dal/backend/primitives/objective_function/logloss_functors_dpc.cpp

```diff
@@ -0,0 +1,532 @@
+        }
+    }
+
+    const Float regularization_factor = L2_;
+
+    const auto kernel_regularization = [=](const Float a, const Float param) {
+        return a + param * regularization_factor;
+    };
+
+    auto add_regularization_event =
+        element_wise(q_, kernel_regularization, out, vec, out, last_iter_deps);
+
+    return add_regularization_event;
+}
+
+template <typename Float>
+sycl::event logloss_hessian_product<Float>::operator()(const ndview<Float, 1>& vec,
+                                                       ndview<Float, 1>& out,
+                                                       const event_vector& deps) {
+    if (fit_intercept_) {
+        return compute_with_fit_intercept(vec, out, deps);
+    }
+    else {
+        return compute_without_fit_intercept(vec, out, deps);
+    }
+}
+
+template <typename Float>
+logloss_function<Float>::logloss_function(sycl::queue& q,
+                                          const table& data,
+                                          const ndview<std::int32_t, 1>& labels,
+                                          Float L2,
+                                          bool fit_intercept,
+                                          std::int64_t bsz)
+        : q_(q),
+          data_(data),
+          labels_(labels),
+          n_(data.get_row_count()),
+          p_(data.get_column_count()),
+          L2_(L2),
+          fit_intercept_(fit_intercept),
+          bsz_(bsz == -1 ? get_block_size(n_, p_) : bsz),
+          hessp_(q, data, L2, fit_intercept, bsz_),
+          dimension_(fit_intercept ? p_ + 1 : p_) {
+    ONEDAL_ASSERT(labels.get_dimension(0) == n_);
+    probabilities_ = ndarray<Float, 1>::empty(q_, { n_ }, sycl::usm::alloc::device);
+    gradient_ = ndarray<Float, 1>::empty(q_, { dimension_ }, sycl::usm::alloc::device);
+    buffer_ = ndarray<Float, 1>::empty(q_, { p_ + 2 }, sycl::usm::alloc::device);
+    if (data_.get_kind() == dal::csr_table::kind()) {
+        sp_handle_.reset(new sparse_matrix_handle(q));
+        set_csr_data(q_, *sp_handle_, static_cast<const csr_table&>(data_));
+    }
+}
+
+template <typename Float>
+logloss_function<Float>::logloss_function(sycl::queue& q,
+                                          comm_t comm,
+                                          const table& data,
+                                          const ndview<std::int32_t, 1>& labels,
+                                          Float L2,
+                                          bool fit_intercept,
+                                          std::int64_t bsz)
+        : q_(q),
+          comm_(comm),
+          data_(data),
+          labels_(labels),
+          n_(data.get_row_count()),
+          p_(data.get_column_count()),
+          L2_(L2),
+          fit_intercept_(fit_intercept),
+          bsz_(bsz == -1 ? get_block_size(n_, p_) : bsz),
+          hessp_(q, comm, data, L2, fit_intercept, bsz_),
+          dimension_(fit_intercept ? p_ + 1 : p_) {
```

Comment:
Pretty much he same logic. Can it be moved to a separate private constructor instead?